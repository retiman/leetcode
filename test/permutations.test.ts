describe('permutations', () => {
  function permutations(xs: Array<number>): Array<Array<number>> {
    if (xs.length === 0) {
      return [[]];
    }

    if (xs.length === 1) {
      return [xs];
    }

    const result = [];

    for (let i = 0; i < xs.length; i += 1) {
      // Get the ith element; we will remove this element from the array, and generate permutations without this
      // element.
      const x = xs[i];

      // Generates all permutations without the ith element.
      const without = [...xs.slice(0, i), ...xs.slice(i + 1, xs.length)];
      const ps = permutations(without);

      // For each of the permutations, put the ith element in front of the permutations, to get all possible
      // permutations.
      ps.forEach(p => p.unshift(x));
      result.push(...ps);
    }

    return result;
  }

  test('run trivial', async () => {
    expect(permutations([])).toStrictEqual([[]]);
    expect(permutations([1])).toStrictEqual([[1]]);
  });

  test('run 1, 2, 3, 4', async () => {
    const ps = permutations([1, 2, 3, 4]);
    const length = 4 * 3 * 2 * 1;

    expect(ps.length).toBe(length);
    expect(ps).toMatchSnapshot();
  });
});
