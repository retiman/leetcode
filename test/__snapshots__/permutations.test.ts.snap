// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`permutations run 1, 2, 3, 4 1`] = `
Array [
  Array [
    1,
    2,
    3,
    4,
  ],
  Array [
    1,
    2,
    4,
    3,
  ],
  Array [
    1,
    3,
    2,
    4,
  ],
  Array [
    1,
    3,
    4,
    2,
  ],
  Array [
    1,
    4,
    2,
    3,
  ],
  Array [
    1,
    4,
    3,
    2,
  ],
  Array [
    2,
    1,
    3,
    4,
  ],
  Array [
    2,
    1,
    4,
    3,
  ],
  Array [
    2,
    3,
    1,
    4,
  ],
  Array [
    2,
    3,
    4,
    1,
  ],
  Array [
    2,
    4,
    1,
    3,
  ],
  Array [
    2,
    4,
    3,
    1,
  ],
  Array [
    3,
    1,
    2,
    4,
  ],
  Array [
    3,
    1,
    4,
    2,
  ],
  Array [
    3,
    2,
    1,
    4,
  ],
  Array [
    3,
    2,
    4,
    1,
  ],
  Array [
    3,
    4,
    1,
    2,
  ],
  Array [
    3,
    4,
    2,
    1,
  ],
  Array [
    4,
    1,
    2,
    3,
  ],
  Array [
    4,
    1,
    3,
    2,
  ],
  Array [
    4,
    2,
    1,
    3,
  ],
  Array [
    4,
    2,
    3,
    1,
  ],
  Array [
    4,
    3,
    1,
    2,
  ],
  Array [
    4,
    3,
    2,
    1,
  ],
]
`;
